src/apps/test/lwaftr.lua

```diff
@@ -0,0 +1,345 @@
+  lib.bitfield(32, ipv6_hdr, 'v_tc_fl', 0, 4, 6) -- IPv6 Version
+  lib.bitfield(32, ipv6_hdr, 'v_tc_fl', 4, 8, 1) -- Traffic class
+  ipv6_hdr.next_header = PROTO_IPV4_ENCAPSULATION
+  ipv6_hdr.hop_limit = DEFAULT_TTL
+  ipv6_hdr.dst_ip = aftr_ipv6
+
+  local ipv6_ipv4_hdr = cast(ipv4_header_ptr_type, ipv6_pkt.data + 14 + ipv6_header_size)
+  ipv6_ipv4_hdr.dst_ip = public_ipv4
+  ipv6_ipv4_hdr.ttl = 15
+  ipv6_ipv4_hdr.ihl_v_tos = C.htons(0x4500) -- v4
+  ipv6_ipv4_hdr.id = 0
+  ipv6_ipv4_hdr.frag_off = 0
+
+  local ipv6_ipv4_udp_hdr, ipv6_payload
+
+  local total_packet_count = 0
+  for _,size in ipairs(conf.sizes) do
+    -- count for IPv4 and IPv6 packets (40 bytes IPv6 encap header)
+    if conf.ipv4_only or conf.ipv6_only then 
+      total_packet_count = total_packet_count + 1
+    else
+      total_packet_count = total_packet_count + 2
+    end
+  end
+
+  ipv6_ipv4_hdr.protocol = 17  -- UDP(17)
+  ipv6_ipv4_udp_hdr = cast(udp_header_ptr_type, ipv6_pkt.data + 34 + ipv6_header_size)
+  ipv6_ipv4_udp_hdr.dst_port = C.htons(12345)
+  ipv6_ipv4_udp_hdr.checksum = 0
+  ipv6_payload = cast(payload_ptr_type, ipv6_pkt.data + 34 + ipv6_header_size + udp_header_size)
+  ipv6_payload.magic = MAGIC
+  ipv6_payload.number = 0
+
+  local o = {
+    b4_ipv6 = b4_ipv6,
+    b4_ipv4 = b4_ipv4,
+    b4_port = conf.b4_port,
+    current_port = conf.b4_port,
+    b4_ipv4_offset = 0,
+    ipv6_address = n_cache_src_ipv6,
+    count = conf.count,
+    current_count = 0,
+    ipv4_pkt = ipv4_pkt,
+    ipv4_hdr = ipv4_hdr,
+    ipv4_payload = ipv4_payload,
+    ipv6_hdr = ipv6_hdr,
+    ipv6_pkt = ipv6_pkt,
+    ipv6_payload = ipv6_payload,
+    ipv6_ipv4_hdr = ipv6_ipv4_hdr,
+    ipv4_udp_hdr = ipv4_udp_hdr,
+    ipv6_ipv4_udp_hdr = ipv6_ipv4_udp_hdr,
+    ipv4_only = conf.ipv4_only,
+    ipv6_only = conf.ipv6_only,
+    protocol = conf.protocol,
+    rate = conf.rate,
+    sizes = conf.sizes,
+    total_packet_count = total_packet_count,
+    bucket_content = conf.rate * 1e6,
+    ipv4_packets = 0, ipv4_bytes = 0,
+    ipv6_packets = 0, ipv6_bytes = 0,
+    ipv4_packet_number = 0, ipv6_packet_number = 0,
+    last_rx_ipv4_packet_number = 0, last_rx_ipv6_packet_number = 0,
+    lost_packets = 0
+  }
+  return setmetatable(o, {__index=Lwaftrgen})
+end
+
+function Lwaftrgen:push ()
+
+  local input = self.input.input
+  local output = self.output.output
+  local ipv6_packets = self.ipv6_packets
+  local ipv6_bytes = self.ipv6_bytes
+  local ipv4_packets = self.ipv4_packets
+  local ipv4_bytes = self.ipv4_bytes
+  local lost_packets = self.lost_packets
+
+  -- count and trach incoming packets
+  for _=1,link.nreadable(input) do
+    local pkt = receive(input)
+    if cast(uint16_ptr_t, pkt.data + OFFSET_ETHERTYPE)[0] == PROTO_IPV6 then
+      ipv6_bytes = ipv6_bytes + pkt.length
+      ipv6_packets = ipv6_packets + 1
+      payload = cast(payload_ptr_type, pkt.data + 34 + ipv6_header_size + udp_header_size)
```

Comment:
Yep. Same again for the ipv4 handling a few lines down. I wonder if it makes more sense to define payload as local outside the loop or if it doesn't matter.